### PR TITLE
Placeholder support for snippet

### PR DESCRIPTION
**Pull request by mg979**

### CODE REVIEW

**Comment by mg979** [comment]
Last commit can be dropped if you don't want to make it configurable (personally I prefer to drop `TODO:` from the placeholders text).

**Comment by danymat** [comment]
Hello, can you take a peak to the changes and if you agree with them ?

**Comment by mg979** [comment]
I don't like the changes to emmylua template, I think one placeholder is enough, a second placeholder can actually get in the way. Sometimes annotations are written with a colon, like:

    ---@param name type: description

You must press `<tab>` one more time to advance a single space, with nothing in between. And if you also want a colon it's even more distracting. Multiple placeholders are ok when there's some text to be skipped between them, otherwise it's just one more keypress.

**Comment by danymat** [comment]
You're right, I reverted the commit.

**Comment by danymat** [comment]
Hello @L3MON4D3, I'm wondering if it's possible with luasnip to highlight the placeholder text with a special highlighting.

This is the current output:

Capture d’écran 2022-03-06 à 14 26 07

I would love to have the `[...]` specially highlighted. We use `lsp_expand` function to expand lsp-styled snippet

**Comment by L3MON4D3** [comment]
Currently not, but I'm actually working on that right now :D
Check L3MON4D3/LuaSnip#325, with these patches you could set the `hl_group` used for eg. placeholders per-snippet.
Although the exact highlight from your Screenshot won't be achievable with Luasnip alone: we can only do one color (hl_group) for all of the placeholder

**Comment by danymat** [comment]
This is exactly what we need ! Thank you ! 

By the way, the existing highlight in the screenshot is not what I meant, this is an automatic highlight for Todo texts in Emmy lua

**Comment by L3MON4D3** [comment]
Alright, that's pretty nice then :D

It's possible to set these `ext_opts` (called that because they are passed to `nvim_buf_set_extmark` as `opts` :D) globally. The `ext_opts` set on the snippet-level are then merged with those (snippet higher prio of course).

This means that a user might want placeholders to be green and text to be blue, but the neogen-generated snippets would have eg. red placeholders, but keep the blue text, which (very much IMO) wouldn't be that nice, either neogens style, and that alone is applied, or only the users.

The best way out (again, IMO :D) in this case would be applying only neogens style to neogens snippets, ignoring global settings.
If you agree (with both the problem and the solution), I'd implement a way to tell luasnip that the snippet-local `ext_opts` shouldn't be merged with the global ones.

**Comment by danymat** [comment]
> The best way out (again, IMO :D) in this case would be applying only neogens style to neogens snippets, ignoring global settings.
If you agree (with both the problem and the solution), I'd implement a way to tell luasnip that the snippet-local ext_opts shouldn't be merged with the global ones.

Yes, in fact I thought it was the default behavior, thanks for explaining. I think that snippet ext_opts should override the global snippet, so that other plugins that want to do like us can achieve highlightings promptly.

**Comment by danymat** [comment]
@axieax I think you will be interested in that

**Comment by axieax** [comment]
Woah this looks exciting! Just wondering if this requires neovim v0.7? I'm trying to test this feature out on 0.6.1 but it doesn't seem to work unfortunately.

**Comment by danymat** [comment]
You have to use snippet integration for this feature to work

**Comment by axieax** [comment]
Ahh thanks for the clarification. Love the new changes! I'm surprised [vim-doge](https://github.com/kkoomen/vim-doge) is able to do this without any snippet integration. Would love to see more snippet engine integrations in the future though (such as [hrsh7th/vim-vsnip](https://github.com/hrsh7th/vim-vsnip)).

**Comment by danymat** [comment]
Technically, it's feasible, I just was too lazy to reinvent the wheel lol. But if a lot of users want internal integration then I will reconsider it

**Comment by L3MON4D3** [comment]
> * Is the `ext_opts` option already active in `lsp_expand` (in PR)

Right, it's not possible to pass the `opts` to the snippet via `lsp_expand` (yet, I'll think about how to best expose that).
What works is
```lua
local types = require("luasnip.util.types")

ls.snip_expand(ls.s("",
	ls.parser.parse_snippet(nil, "a ${1:snippet} ${2:with placeholders}"),
	{
		child_ext_opts = {
			-- for highlighting the placeholders
			[types.insertNode] = {
				-- when inside placeholder
				active = {
					hl_group = "GruvboxRed"
				},
				-- when outside placeholder, but in snippet
				passive = {
					hl_group = "GruvboxOrange"
				},
				-- when outside snippet
				snippet_passive = {
					hl_group = "GruvboxYellow"
				}
			}
		},
		-- prevent mixing styles
		merge_child_ext_opts = false
	}
))
```

File: lua/neogen/snippet.lua [hunk 1]
```diff
@@ -52,13 +52,45 @@ end
...
+    local _snip = table.concat(snip, "\n")
+
+    ls.snip_expand(
+        ls.s("", ls.parser.parse_snippet(nil, _snip), {
+
+            child_ext_opts = {
+                -- for highlighting the placeholders
+                [types.insertNode] = {
+                    -- when inside placeholder
+                    active = {
```
Comment (L3MON4D3) [inline lua/neogen/snippet.lua:76]: `active` doesn't need to be set here explicitly, it inherits the values from `passive` (and `passive` in turn from `snippet_passive`) (just fyi, I'd understand having it there for clarity)

Comment (danymat) [reply lua/neogen/snippet.lua]: Oh thanks ! Removed it.

**Comment by danymat** [comment]
Now that we have highlights in luasnip, next step is to create neogen highlights. (except if you guys have some default highlights that would want to use, like Todo ? which I'm not even sure this one is default)

**Comment by danymat** [comment]
Hello everyone, I feel like this PR is okay to be merged ! very excited for you guys to try out ! 🥳

**Pull request merged**